README
This is the README file for the Willows TWIN library source
distribution.  For instructions on how to install the TWIN libraries
and the emulator, please see the INSTALL file.  For a list of changes
since the last release, please see the ChangeLog file.  This is the
official release of the Willows TWIN Library version 3.1.14.

What is Willows TWIN?

Willows TWIN is a GNU software package for emulating the Microsoft
Win32 API.  It consists of a library set and an emulator program.  The
emulator program allows the execution of Windows binary applications
on supported platforms; using the native Intel x86 processors, or
through the built-in instruction set interpreter.  The native
libraries allow programmers to natively build Win32 and MFC
applications from source code.

Because Willows TWIN libraries are an implementation of the Win32
API, certain operating system features are required to build these
libraries.  The following is a list of required tools and features
for building TWIN.

	X Window System version 11
	Shared library support (dynamic library loading)
	GCC 2.7.2 or later
	Flex
	Bison
	GNU Make

The following processors are supported:

	Intel x86
	Sparc
	Alpha [1]
	PowerPC [1]

The following operating systems are supported;

	GNU/Linux
	Solaris 2.5
	HPUX 11.0 [1]
	AIX 4.3.1 [1]

[1] Not tested by maintainer.

The goal of Willows TWIN library is to allow portation of Windows
applications and execution of existing Windows binaries.


Major Changes Since Last Release

Though it has been a while since the previous release.  This release
should be considered a bug fix release only.  The following are the
major points of changes:

20010428
	o Mostly code cleanup.

19990529
	o New interpreter code.  This code base is shared with the
	  DOSEMU project.

19990407
	o Directory hierarchy now mirrors that of WINE.
	o Windows header files renamed to match spec.
	o Build from a different location than the source directory.
	o Sample applications are now re-integrated back into the
	  distribution, for debugging purposes.


Open Development model

The Willows TWIN libraries development has completely adopted the open
(bazaar) development model.  The latest development sources are
available via anonymous CVS server.  Instructions for accessing the
CVS tree can be obtained from;

	http://www.willows.com/

For the impatient, the following command can be used.

	cvs -d :pserver:cvsguest@cvs.willows.com:/cVs login

Use 'twin', without qoutes, as the password.

	cvs -d :pserver:cvsguest@cvs.willows.com:/cVs checkout -P twin

The '-P' is needed because of the directory changes since 3.1.12.  To
get the latest updates, type

	cvs -q update -P -d

from inside the 'twin' directory.

Snapshots will also be created as often as possible, depending on the
amount of contributions (changes to the cvs tree).  A mailing list is
also provided for developer discussions and user assistance.  The
Willows web page has further information about these mailing lists and
archives.


Last Notes

If anyone is interested in contributing to Willows TWIN libraries
development, please see the TODO file.  We are in need of help.  You
don't need to know how to program.  There are lots of other tasks that
need to be done.

Thank you for using the Willows TWIN libraries software.


ChangeLog
--------------------------------------------------------------------------------
2001-04-29  Jimen Ching  <jching@core2>

	* README, windows/Release.h: Increase version number.

2001-04-27  Jimen Ching  <jching@core2>

	* include/winbase.h, windows/File32.c:
	MFC wants HFILE to be an integral type.

	* include/commctrl.h, include/prsht.h: Add new flags.

2001-01-28  Jimen Ching  <jching@core2>

	* intp32/fp87.c: Fix floating point load/store on non-x86 systems.

2001-01-17  Jimen Ching  <jching@core2>

	* samples/bn2demo/binlib.c, samples/cliptext/cliptext.c, samples/ctrldemo/proppage.c, samples/dlldemo/dlldemo.h, samples/mdidemo/mdidemo.c, samples/mdidemo/mdidemo.h, samples/msgdemo/msgdemo.c, samples/bindemo/bindemo.c, samples/bindemo2/binlib2.c:
	Clean up compile warnings.

	* include/commctrl.h: Add prototype for generic use.

2001-01-16  Jimen Ching  <jching@core2>

	* windows/MessageBox.c, windows/Messages.c, windows/Meta.h, windows/MetaObjects.c, windows/Metafile.c, windows/Metaplay.c, windows/Module.c, windows/PrintManager.c, windows/PrinterGraphics.c, windows/Resources.c, windows/Time.c, windows/WBmp.c, windows/WBmp.h, windows/WImageIO.c, windows/WSysMenu.c, windows/WinMain.c, windows/WindowBytes.c, windows/WindowCreate.c, windows/WindowDC.c, windows/WindowPos.c, windows/WindowProps.c, windows/WindowRedraw.c, windows/lib32.c, if1632/PrintIF.c, if1632/SpecialIF.c, include/Resources.h, include/toolhelp.h, include/winbase.h, include/windef.h, include/windows.h, include/wingdi.h, memory/heap.c, msdos/dos_findfile.c, multimedia/midi.c, multimedia/wave.c, ole/OleSvrIF.c, scheduler/Critical.c, win32/LoadPE.c, win32/Resources32.c, windows/Clipboard.h, windows/DefWnd.c, windows/Dialog.c, windows/DisplayGraphics.c, windows/File32.c, windows/FileSys.c, windows/KrnFileIO.c, windows/LoadDLL.c, windows/MDI.c, windows/MenuAPI.c, windows/MenuProc.c, controls/EditMisc.c, controls/Listbox.c, controls/Static.c, debugger/dbg.y, debugger/debugger.h, dlls/commctrl/WStatusb.c, dlls/commctrl/WToolbar.c, dlls/commdlg/ChooseFont.c, dlls/commdlg/CommdlgIF.c, dlls/commdlg/PrintDlg.c, dlls/ddeml/DdeML.c, dlls/lzexpand/LZExpand.c, dlls/shell/About.c, dlls/shell/Shell.c, dlls/shell/ShellIF.c, dlls/toolhelp/Task.c, dlls/toolhelp/ToolHelp.c, dlls/toolhelp/ToolHelpIF.c, dlls/winsock/WinSock.c, dlls/winsock/WinSockIF.c, graphics/GdiImages.c, graphics/GdiObjects.c, graphics/x11drv/DrvIPC.c, graphics/x11drv/DrvWindows.c, graphics/x11drv/XEvents.c, graphics/x11drv/XICCCM.c, if1632/Callback.c, if1632/CommIF.c, if1632/ConvertMsg.c, if1632/GenericIF.c, if1632/HookIF.c, if1632/MemoryIF.c, if1632/OEMRoutines.c, controls/Button.c, controls/ButtonDraw.c, controls/ComboBox.c, controls/Edit.c:
	Change handle types to pointers and cleaned up all warnings.

	* include/print.h, include/mmsystem.h:
	Make prototype consistent with definition.

	* include/io.h: Add prototypes for osfhandle APIs.

	* dlls/commctrl/WTlbCust.c, dlls/commctrl/WTView.c, dlls/commctrl/WTVLBox.c, dlls/commctrl/WProgBar.c:
	Used wrong cast.

2001-01-14  Jimen Ching  <jching@core2>

	* configure, configure.in: Add w32sys.dll.

	* dlls/Makefile.in: Add w32sys.

	* dlls/w32sys/Makefile.in, dlls/w32sys/w32sys.c, dlls/w32sys/w32sys.def, dlls/w32sys/w32sys.rc, dlls/w32sys/w32sysBin.c:
	Add w32sys.dll support.  Some binary apps do depend on it.

	* msdos/mfs_fileio.c:
	Fix problem where mtime was modified when the file was fclose'ed.

2001-01-13  Jimen Ching  <jching@core2>

	* dlls/winsock/WinSock.def: Undo changes in 1.2.

2001-01-07  Jimen Ching  <jching@core2>

	* include/winmapclib.h: Additional C library API.

	* graphics/x11drv/DrvText.c: Remove prototype for ctype API.

	* mfc/Makefile.in:
	Include -fno-for-scope because MFC still uses pre-ANSI C++ constructs.

2001-01-01  Jimen Ching  <jching@core2>

	* debugger/debugger.h, include/thread.h, include/winmaptypes.h, windows/Process32.h:
	Remove the use of BOOL32 in TWIN proper.

	* include/windef.h, include/windows.h:
	Move typedefs from windows.h to windef.h

	* include/winbase.h: Removed macros that were moved into winerror.h.

2000-12-31  Jimen Ching  <jching@core2>

	* tools/rc/rclex.l: Use correct free function.

2000-12-23  Jimen Ching  <jching@core2>

	* graphics/GdiBitmaps.c: Twine merge: support new size.

	* graphics/x11drv/DrvGraphics.c:
	Twine merge: be more specific on styles mask.

	* graphics/x11drv/DrvImages.c: Twine merge: use default mask.

2000-12-17  Jimen Ching  <jching@core2>

	* graphics/x11drv/DrvWindows.c:
	Twine merge: immediate refresh after raise.

	* documentation/lib-config.txt, documentation/readme.twinrc, documentation/win.ini:
	Change FontSubstitutes to Font Alias.

	* graphics/x11drv/DrvText.c: Cleanup logstr.
	Changed FontSubstitutes to Font Alias.

2000-12-10  Jimen Ching  <jching@core2>

	* dlls/commdlg/GetOpenFileName.c: Twine merge: post result explicitly.

	* debugger/dbg.y, debugger/debugger.h, debugger/expr.c, debugger/glue.c, debugger/registers.c, debugger/sig_context.h, debugger/types.c:
	Fixed symbol collision.  Removed unused code.

	* windows/ExecSysV.c: Use defined API.

	* dlls/toolhelp/Info.c:
	TWINE merge: exit TWIN program if TerminateApp is called.

	* dlls/winsock/WinSock.def: Make consistent with other def files.

2000-12-04  Jimen Ching  <jching@core2>

	* controls/Edit.c, controls/Edit.h: Added dwExStyle to Edit control.

	* controls/Button.c, controls/ComboBox.c, controls/Frame.c, controls/Listbox.c, controls/Scroll.c, controls/Static.c:
	Added CALLBACK attribute to window procedure definitions.

2000-09-19  Jimen Ching  <jching@core2>

	* if1632/SpecialIF.c, include/Net.h, windows/Net.c:
	Update API signature.

	* windows/Profile.c: When flushing cache, no file is specified.

2000-09-04  Jimen Ching  <jching@core2>

	* windows/KrnTask.c, windows/LoadLibrary.c:
	Add FreeModule to ExitTask, so apps can be re-entered.

	* windows/ObjEngine.c: Fix apidepth sync.

2000-09-02  Jimen Ching  <jching@core2>


tools/lzc.c
/*    
	lzc.c	2.5
    	Copyright 1997 Willows Software, Inc. 

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; see the file COPYING.LIB.  If
not, write to the Free Software Foundation, Inc., 675 Mass Ave,
Cambridge, MA 02139, USA.


For more information about the Willows Twin Libraries.

	http://www.willows.com	

To send email to the maintainer of the Willows Twin Libraries.

	mailto:twin@willows.com 

 */


#include "windows.h"
#include "kerndef.h"

#include <sys/types.h>
#include <sys/stat.h>

/*
 * clone of compress/expand from microsoft SDK
 *
 *	single source for both compression and expansion 
 *	cc -o lzc          lzc.c 	to get compression program
 *	cc -o lze -DEXPAND lzc.c	to get expansion program
 *
 *	missing the prompting for input and output files 
 *	when not given on the command line...
 *
 *	does not verify the file size as given by the header
 *	after expanding the file
 *
 *	does not validate that the file is actually compressed
 *	by reading the 4 byte signature
 *
 *	does not support the -r option to automatically name the
 *	output file for compression
 *
 *	does not support multiple file input to a directory output
 *	for compression
 *
 ****************************************************************
 *
 *	still don't know all the fields in the 14 byte header
 *	as far as I can tell
 *
 *	0-3 	are SZDD some kind of signature 
 *	4-9	are 6 bytes of ????
 *	10-13   are the file size 
 *
 */
 
#ifndef EXPAND
/* define the strings when compiled for compression */
char *szexestr = "Compression";		/* what type program */
char *szopstr  = "Compressing";		/* what its doing    */
char *szopcmd  = "compresses";		/* what it does      */
#define LZ	Encode

#else

/* define the strings when compiled for expansion */
char *szexestr = "Expansion";
char *szopstr  = "Expanding";
char *szopcmd  = "expands";
#define LZ	Decode
#endif

/*****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define N		4096	/* size of ring buffer */
#define F		16	/* 18 upper limit for match_length */
#define THRESHOLD	2       /* minimum character length to compress */
#define NIL		N	/* index for root of binary search trees */

unsigned long 
		textsize = 0,	/* text size counter */
		codesize = 0;	/* code size counter */
/*********************************************************************
   ring buffer of size N, with extra F-1 bytes to facilitate string 
   comparison of longest match.  These are set by InsertNode()
**********************************************************************/
BYTE 	text_buf[N + F - 1];	

/**********************************************************************
	left & right children & parents
***********************************************************************/
int	match_position, match_length,  
	lson[N + 1], rson[N + 257], dad[N + 1];  

/**********************************************************************
	input & output files
***********************************************************************/
FILE	*infile, *outfile;  

/**********************************************************************
	global name of source file
***********************************************************************/
char	*sourcefile;

/**********************************************************************/
/* this is the header stuck to the front of a compressed file 	      */
/**********************************************************************/
char header[] = {
	'\x53', '\x5A', '\x44', '\x44', 
	'\x88', '\xF0', '\x27', '\x33', '\x41', '\x00'
};

void InitTree(void)  /* initialize trees */
{
	int  i;

	/* For i = 0 to N - 1, rson[i] and lson[i] will be the right and
	   left children of node i.  These nodes need not be initialized.
	   Also, dad[i] is the parent of node i.  These are initialized to
	   NIL (= N), which stands for 'not used.'
	   For i = 0 to 255, rson[N + i + 1] is the root of the tree
	   for strings that begin with character i.  These are initialized
	   to NIL.  Note there are 256 trees. */

	for (i = N + 1; i <= N + 256; i++) rson[i] = NIL;
	for (i = 0; i < N; i++) dad[i] = NIL;
}

void InsertNode(int r)
	/* Inserts string of length F, text_buf[r..r+F-1], into one of the
	   trees (text_buf[r]'th tree) and returns the longest-match position
	   and length via the global variables match_position and match_length.
	   If match_length = F, then removes the old node in favor of the new
	   one, because the old one will be deleted sooner.
	   Note r plays double role, as tree node and position in buffer. */
{
	int  i, p, cmp;
	BYTE  *key;

	cmp = 1;  key = &text_buf[r];  p = N + 1 + key[0];
	rson[r] = lson[r] = NIL;  match_length = 0;
	for ( ; ; ) {
		if (cmp >= 0) {
			if (rson[p] != NIL) p = rson[p];
			else {  rson[p] = r;  dad[r] = p;  return;  }
		} else {
			if (lson[p] != NIL) p = lson[p];
			else {  lson[p] = r;  dad[r] = p;  return;  }
		}
		for (i = 1; i < F; i++)
			if ((cmp = key[i] - text_buf[p + i]) != 0)  break;
		if (i > match_length) {
			match_position = p;
			if ((match_length = i) >= F)  break;
		}
	}
	dad[r] = dad[p];  lson[r] = lson[p];  rson[r] = rson[p];
	dad[lson[p]] = r;  dad[rson[p]] = r;
	if (rson[dad[p]] == p) rson[dad[p]] = r;
	else                   lson[dad[p]] = r;
	dad[p] = NIL;  /* remove p */
}

void DeleteNode(int p)  /* deletes node p from tree */
{
	int  q;
	
	if (dad[p] == NIL) return;  /* not in tree */
	if (rson[p] == NIL) q = lson[p];
	else if (lson[p] == NIL) q = rson[p];
	else {
		q = lson[p];
		if (rson[q] != NIL) {
			do {  q = rson[q];  } while (rson[q] != NIL);
			rson[dad[q]] = lson[q];  dad[lson[q]] = dad[q];
			lson[q] = lson[p];  dad[lson[p]] = q;
		}
		rson[q] = rson[p];  dad[rson[p]] = q;
	}
	dad[q] = dad[p];
	if (rson[dad[p]] == p) rson[dad[p]] = q;  else lson[dad[p]] = q;
	dad[p] = NIL;
}

void Encode(void)
{
	int  i, c, len, r, s, last_match_length, code_buf_ptr;
	BYTE  code_buf[17], mask;
	struct stat buf;
	unsigned long size;
	
	for(i=0;i<10;i++) {
		putc(header[i],outfile);
	}

	/* now put out four bytes of size */
	stat(sourcefile,&buf);	
	size = buf.st_size;
	putc((size    )&0xff,outfile);
	putc((size>> 8)&0xff,outfile);
	putc((size>>16)&0xff,outfile);
	putc((size>>24)&0xff,outfile);
	
	InitTree();  /* initialize trees */

/***************************************************************************
 	code_buf[1..16] saves eight units of code, and
	code_buf[0] works as eight flags, "1" representing that the unit
	is an unencoded letter (1 byte), "0" a position-and-length pair
	(2 bytes).  Thus, eight units require at most 16 bytes of code. 
****************************************************************************/
		
	code_buf[0] = 0;  

	code_buf_ptr = mask = 1;

	s = 0;  r = N - F;
	/* Clear the buffer with any character that will appear often. */
	for (i = s; i < r; i++) text_buf[i] = ' ';  

	/* Read F bytes into the last F bytes of the buffer */
	for (len = 0; len < F && (c = getc(infile)) != EOF; len++)
		text_buf[r + len] = c;  

	/* text of size zero */
	if ((textsize = len) == 0) 
		return;  

	/****************************************************************** 
	Insert the F strings, each of which begins with one or more 'space' 
	characters.  Note the order in which these strings are inserted.  
	This way, degenerate trees will be less likely to occur. 
	********************************************************************/
	for (i = 1; i <= F; i++) 
		InsertNode(r - i);  

	/**********************************************************
	Finally, insert the whole string just read.  The
	global variables match_length and match_position are set. 
 	***********************************************************/
	InsertNode(r);  

	do {
		/* match_length may be spuriously long near the end of text. */
		if (match_length > len) 
			match_length = len;  

		/* Not long enough match.  Send one byte. */
		if (match_length <= THRESHOLD) {
			/* Send uncoded. */
			match_length = 1;  
			code_buf[0] |= mask;  /* 'send one byte' flag */
			code_buf[code_buf_ptr++] = text_buf[r];  
		} else {
			/************************************
			Send position and length pair. 
			Note match_length > THRESHOLD. 
			*************************************/
			code_buf[code_buf_ptr++] = (BYTE) match_position;
			code_buf[code_buf_ptr++] = 
				(BYTE) (((match_position >> 4) & 0xf0)
			  	| (match_length - (THRESHOLD + 1)));  
		}

		/* Shift mask left one bit. */
		if ((mask <<= 1) == 0) {  
			/* Send at most 8 units of code together */
			for (i = 0; i < code_buf_ptr; i++)  
				putc(code_buf[i], outfile); 

			codesize += code_buf_ptr;
			code_buf[0] = 0;  
			code_buf_ptr = mask = 1;
		}
		last_match_length = match_length;
		for (i = 0; i < last_match_length &&
				(c = getc(infile)) != EOF; i++) {
			DeleteNode(s);		/* Delete old strings and */
			text_buf[s] = c;	/* read new bytes */
			if (s < F - 1) text_buf[s + N] = c;  

			/**************************************************
			 If the position is near the end of buffer, extend 
			   the buffer to make string comparison easier. 
			***************************************************/

			s = (s + 1) & (N - 1);  r = (r + 1) & (N - 1);
				/* Since this is a ring buffer, increment the 
				   position modulo N. */

			/* Register the string in text_buf[r..r+F-1] */
			InsertNode(r);	
		}

		/* After the end of text, */
		while (i++ < last_match_length) {	
			/* no need to read, but */
			DeleteNode(s);					
			s = (s + 1) & (N - 1);  r = (r + 1) & (N - 1);

			/* buffer may not be empty. */
			if (--len) InsertNode(r);		
		}
	} while (len > 0);	
	/* until length of string to be processed is zero */

	/* Send remaining code. */
	if (code_buf_ptr > 1) {		
		for (i = 0; i < code_buf_ptr; i++) 
			putc(code_buf[i], outfile);
		codesize += code_buf_ptr;
	}

	codesize += 14;

	printf("%s: %ld bytes compressed to %ld bytes,",
		sourcefile, buf.st_size, codesize);

	printf(" %d%% savings.\n",
		100 - ((100*codesize) / buf.st_size));
}

void Decode(void)
{
	int  i, j, k, r, c;
	int  ii,jj;
	unsigned int  flags;
	int	phase;
	
	/* skip the header... */
	for (i=0;i<14;i++)
		c = getc(infile);

	for (i = 0; i < N - F; i++) text_buf[i] = ' ';
	r = N - F;  flags = 0;

	for ( ; ; ) {
		phase = 0;			
		if (((flags >>= 1) & 0x100) == 0) {
			if ((c = getc(infile)) == EOF) 
				break;

			/* uses higher byte cleverly to count eight */
			flags = c | 0xff00;		
		}							

		if (flags & 1) {
			phase = 1;
			if ((c = getc(infile)) == EOF) 
				break;

			putc(c, outfile);  
			text_buf[r++] = c;  r &= (N - 1);
		} else {
			phase = 2;
			if ((i = getc(infile)) == EOF) break;
			phase = 3;
			if ((j = getc(infile)) == EOF) break;

			ii = i;
			jj = j;

			i |= ((j & 0xf0) << 4);  j = (j & 0x0f) + THRESHOLD;

			for (k = 0; k <= j; k++) {
				c = text_buf[(i + k) & (N - 1)];

				putc(c, outfile);  
				text_buf[r++] = c;  r &= (N - 1);
			}
		}
	}
}

int main(int argc, char *argv[])
{
	char  *s;
	int i;
	
	if (argc != 3) {
		printf("%s:file1 file2 [%s file1 into file2].\n",
			argv[0],szopcmd);
		return EXIT_FAILURE;
	}
	if ((s = argv[1], (infile  = fopen(s, "rb")) == NULL)
	 || (s = argv[2], (outfile = fopen(s, "wb")) == NULL)) {
		printf("??? %s\n", s);  return EXIT_FAILURE;
	}

	printf("Willows Software (R) File %s Utility %s\n", 
		szexestr,"Version 1.00");

	printf("Copyright (C) %s %s All rights reserved.\n\n",
		"Willows Software",
		"1990-1992.");
	printf("%s %s to %s.\n",
		szopstr,argv[1],argv[2]);
	sourcefile = argv[1];

	LZ();  
	
	fclose(infile);  fclose(outfile);
	return EXIT_SUCCESS;
}
